grpc/shared_conn.go

```diff
@@ -319,3 +325,107 @@ func (sc *SharedConn) Close() error {
 
 	return multierr.Combine(err, sc.grpcConn.Close())
 }
+
+// NewLocalPeerConnection creates a peer connection that only accepts loopback
+// address candidates.
+func NewLocalPeerConnection(logger golog.Logger) (*webrtc.PeerConnection, error) {
+	m := webrtc.MediaEngine{}
+	if err := m.RegisterDefaultCodecs(); err != nil {
+		return nil, err
+	}
+	i := interceptor.Registry{}
+	if err := webrtc.RegisterDefaultInterceptors(&m, &i); err != nil {
+		return nil, err
+	}
+
+	var settingEngine webrtc.SettingEngine
+	settingEngine.SetIncludeLoopbackCandidate(true)
+	settingEngine.SetRelayAcceptanceMinWait(3 * time.Second)
+	settingEngine.SetIPFilter(func(ip net.IP) bool {
+		// Disallow ipv6 addresses since grpc-go does not currently support IPv6 scoped literals.
+		// See related grpc-go issue: https://github.com/grpc/grpc-go/issues/3272.
+		//
+		// Stolen from net/ip.go, `IP.String` method.
+		// Disallow non loopback addresses as this is a local peer connection
+		if p4 := ip.To4(); len(p4) == net.IPv4len && p4.IsLoopback() {
+			logger.Debugf("SetIPFilter allowing loppback ip: %s", ip.String())
+			return true
+		}
+		logger.Debugf("SetIPFilter disallowing ip: %s", ip.String())
+
+		return false
+	})
+
+	options := []func(a *webrtc.API){webrtc.WithMediaEngine(&m), webrtc.WithInterceptorRegistry(&i)}
+	if utils.Debug {
+		settingEngine.LoggerFactory = WebRTCLoggerFactory{logger}
+	}
+	options = append(options, webrtc.WithSettingEngine(settingEngine))
+	webAPI := webrtc.NewAPI(options...)
+
+	// attempt to construct a PeerConnection
+	pc, err := webAPI.NewPeerConnection(webrtc.Configuration{})
+	if err != nil {
+		logger.Debugw("Unable to create optional peer connection for module. Skipping WebRTC for module...", "err", err)
+		return nil, err
+	}
+	return pc, nil
+}
+
+// WebRTCLoggerFactory wraps a golog.Logger for use with pion's webrtc logging system.
+type WebRTCLoggerFactory struct {
+	Logger golog.Logger
```

Comment:
I'm guessing this was just copied from goutils for expediency? That's fine if so. I can dabble in figuring out what our longer-term options are here.

Reply:
yep correct